File: clojure/armstrong-numbers/src/armstrong_numbers.clj
(ns armstrong-numbers)

(defn- expt [x n]
  (reduce * (repeat n x)))

(defn- number->digits [n]
  (if (zero? n)
    ()
    (cons (mod n 10)
          (number->digits (quot n 10)))))

(defn armstrong? [num]
  (let [digits       (number->digits num)
        digits-count (count digits)
        armstrong    (->> digits
                          (map #(expt % digits-count))
                          (reduce +))]
    (= num armstrong)))

File: haskell/armstrong-numbers/src/ArmstrongNumbers.hs
module ArmstrongNumbers (armstrong) where

armstrong :: Integral a => a -> Bool
armstrong n = n == sumOfDigits
  where
    digits = intToDigits n
    sumOfDigits = sum $ map (^ length digits) digits

intToDigits :: Integral a => a -> [a]
intToDigits 0 = []
intToDigits x = x `mod` 10 : intToDigits (x `div` 10)
